### PR TITLE
Powershell scripts

### PR DESCRIPTION
**Pull request by blevkovych**
Created PowerShell scripts using which you can authorize to back end of application and get pet by its Id also exporting bearer token along the way.

### CODE REVIEW

File: scripts/getpet.ps1 [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+try
+{
+    $response = Invoke-RestMethod -Uri $server/api/pets/$pet_id -Headers $headers -ContentType "application/json"
+}
+catch
+{  
+    if (  $_.Exception.response.StatusCode -eq "Unauthorized" )
+    {
+        Write-Output "Token is expired, authorize to generate new one!"
```
Comment (omelianlevkovych) [inline scripts/getpet.ps1]: Lets change Write-Output to Write-Warning so the user will not miss it.

File: scripts/auth.ps1 [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+try
+{
+    $response = Invoke-WebRequest -Uri $server/api/account/login -Method POST -Body $body -Headers $headers
+}
+catch
+{
+    if (  $_.Exception.response.StatusCode -eq "Unauthorized" )
+    {
+        Write-Output "Incorrect username or password"
```
Comment (omelianlevkovych) [inline scripts/auth.ps1]: Write-Error

**Comment by omelianlevkovych** [comment]
1. It would be nice to have the specific directory 'PS_scripts' (you can suggest better naming for it) inside parent scripts directory, so that we can distinguish it from other scripts which could be added later on.
2. I do not like the way you interact within scripts by making them call each other. I would strongly recommend you to create one main.ps1 script and include all other into that one. You can achieve this by setting the external scripts with adding following line - . ".\auth.ps1".
The 'not main' scripts should have just functions, using which you can provide some sort of PowerShell API for our solution. 
The main reasons for this are: 
- It will make our scripts less coupled to each other which will improve maintainability, future possible testing, readability of the code.
- It will allow future scripts to use already existing one (PowerShell API) without writing unnecessary code.
- As our solution grows we can easily extend some of the functions (again lets call it PowerShell Application Programming Interface) just by changing it in one place.

File: scripts/auth.ps1 [hunk 3]
```diff
@@ -0,0 +1,46 @@
+param (
+    [string]$server = "https://localhost:5001"
+)
+
+# Ignoring self-signed certificates
+add-type @"
```
Comment (omelianlevkovych) [inline scripts/auth.ps1]: Hah, very clever solution, however I am strongly against with using it, especially in our prod environment (AWS).
First of all, it can lead some security risks which is very dangerous and should be strongly forbidden.
Secondly, as we can see, you use such trick in other scripts as well. I can assume that default settings will still remain.
I admit that did not have much time to help you with SSL due to bug fixing client side, my work and other projects, however first thing I will do is looking for the right solution to unblock you. Cheers!

File: scripts/auth.ps1 [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    using System.Security.Cryptography.X509Certificates;
+    public class TrustAllCertsPolicy : ICertificatePolicy {
+        public bool CheckValidationResult(
+            ServicePoint srvPoint, X509Certificate certificate,
+            WebRequest request, int certificateProblem) {
+            return true;
+        }
+    }
+"@
+[System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy
```
Comment (omelianlevkovych) [inline scripts/auth.ps1]: This one also should not be commited!

File: scripts/menu.ps1 [hunk 5]
```diff
@@ -0,0 +1,22 @@
+param (
+    [string]$server = "https://localhost:5001"
+)
+$item = Read-Host @"
+1: Get pet by id.
+2: Unavailable
+3: Unavailable
+Please select the task [1-3] 
+"@
+switch ($item)
```
Comment (omelianlevkovych) [inline scripts/menu.ps1]: switch should also have default behavior

File: scripts/auth.ps1 [hunk 6]
```diff
@@ -0,0 +1,46 @@
+param (
```
Comment (omelianlevkovych) [inline scripts/auth.ps1]: Please, add Set-StrictMode to your scripts. After refactoring, when you will provide PowerShell API with necessary functions, I believe that we can use it only once in main.ps1

File: scripts/auth.ps1 [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+[System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy
+
+$username = Read-Host 'Please enter your username'
+$password = Read-Host 'Please enter your password' -AsSecureString
+$body = @{
+    Username=$username
+    Password=[Runtime.InteropServices.Marshal]::PtrToStringAuto([Runtime.InteropServices.Marshal]::SecureStringToBSTR($password))
+} | ConvertTo-Json
+$headers = @{
+    "Content-Type" = "application/json"
```
Comment (omelianlevkovych) [inline scripts/auth.ps1]: The PowerShell code convention insist on using single quotas instead of double quotas if there are no escape sequences or variables. The reason for this is that single quotas disable the expansion of variables, which will reduce the unexpected results.

File: scripts/auth.ps1 [hunk 8]
```diff
@@ -0,0 +1,46 @@
+param (
+    [string]$server = "https://localhost:5001"
```
Comment (omelianlevkovych) [inline scripts/auth.ps1]: It would be nice to include Update-Help script line in the beginning of script scope code

File: scripts/auth.ps1 (hunk 6 above)
Comment (omelianlevkovych) [reply scripts/auth.ps1]: Also it would be nice to have #Requries statement for the PowerShell version, which you have used to create the script. Except version you can also force the user to use administrator rights if needed and do other cool things.

File: scripts/auth.ps1 (hunk 2 above)
Comment (blevkovych) [reply scripts/auth.ps1]: In my opinion Write-Warning will better look in this case.
image

Comment (omelianlevkovych) [reply scripts/auth.ps1]: Sure, lets do it this way! 👍

File: scripts/auth.ps1 (hunk 3 above)
Comment (blevkovych) [reply scripts/auth.ps1]: I can remove this block of code, but then this scripts will become useless until you get a trusted certificate from CA. My suggestion is remove this block only after you get a certificate.

File: scripts/auth.ps1 (hunk 4 above)
Comment (blevkovych) [reply scripts/auth.ps1]: Remove only after certificate.

**Comment by blevkovych** [comment]
Please review new changes to the scripts.

File: scripts/auth.ps1 (hunk 8 above)
Comment (blevkovych) [reply scripts/auth.ps1]: In my opinion it's unnecessary, misleading and out of scope. It can cause errors if you don't have administrator privileges.

File: scripts/auth.ps1 (hunk 3 above)
Comment (omelianlevkovych) [reply scripts/auth.ps1]: Agree, lets do this way.

File: scripts/auth.ps1 (hunk 8 above)
Comment (omelianlevkovych) [reply scripts/auth.ps1]: The Update-Help is nice to have feature. Lets add than possibility for user to choose, does he actually want to update it?

**Comment by omelianlevkovych** [comment]
Please, do not write standalone comments asking for review the code again. Such comment should be related mostly to the already existing comments that the PR reviewer has already created. You can always just write down under some of the existing comments a short answer, such as 'Thanks, done' etc. It will provide the better traceability what has been changed. The standalone comments is okay to have mostly in case they are not relevant to already existing comments, you can always ask some additional questions, or provide the new topic for the discussion. Thanks.

File: scripts/getpet.ps1 (hunk 1 above)
Comment (blevkovych) [reply scripts/getpet.ps1]: Done

File: scripts/menu.ps1 (hunk 5 above)
Comment (blevkovych) [reply scripts/menu.ps1]: Done.

File: scripts/auth.ps1 (hunk 6 above)
Comment (blevkovych) [reply scripts/auth.ps1]: D O N E

File: scripts/auth.ps1 (hunk 7 above)
Comment (blevkovych) [reply scripts/auth.ps1]: Changed, but one more check would be appreciated. + Could you please elaborate where I should and shouldn't use single, double quotas.

File: scripts/Auth_scripts/main.ps1 [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# Requires PSVersion 5.1.18362.1171
+. ".\auth.ps1"
+. ".\getpet.ps1"
+. ".\menu.ps1"
+
+Set-StrictMode -version 1.0
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/main.ps1]: Your just above comment says that this script requires 5.1 version, why do we use strict mode v1.0?

File: scripts/Auth_scripts/main.ps1 [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        public bool CheckValidationResult(
+            ServicePoint srvPoint, X509Certificate certificate,
+            WebRequest request, int certificateProblem) {
+            return true;
+        }
+    }
+"@
+[System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy
+
+$url = "https://localhost:5001"
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/main.ps1]: Lets rename it as urlDomain? It is fine to leave it as hard coded for now, but we probably will have to let user to input it via CLI later on (when we will have some part of project in another domain also after deploying to prod). You can change it now, if you want to.

Comment (omelianlevkovych) [reply scripts/Auth_scripts/main.ps1]: Okay I see how you do it right after this line, lets leave it for now hardcoded. Still it would be nice to rename the variable.

File: scripts/Auth_scripts/auth.ps1 [hunk 11]
```diff
@@ -0,0 +1,36 @@
+Function authUser($url)
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/auth.ps1]: Lets declare the exact type for the $url variable - function AuthUser([string]$urlDomain)
Please, rename your public functions to start with capital letter.

File: scripts/Auth_scripts/auth.ps1 [hunk 12]
```diff
@@ -0,0 +1,36 @@
+Function authUser($url)
+{
+    while($true)
+    {
+        $username = Read-Host 'Please enter your username'
+        $password = Read-Host 'Please enter your password' -AsSecureString
+        $body = @{
+            Username=$username
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/auth.ps1]: Put the whitespace between the assigning symbol =

File: scripts/Auth_scripts/auth.ps1 [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            }
+            else 
+            {
+               throw
+            }
+        }
+        break
+    }
+    $token = (($response.content | ConvertFrom-Json).token).ToString()
+    $env:TOKEN = $token
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/auth.ps1]: Do not store any not operating system relevant variables in the environment variables! Instead, you can use global variables in your scripts mostly. Here is an example using cmdlet: Set-Variable -Name "accessToken" -Scope global -Value $accessToken. 
Please, rename the token variable to the accessToken, so it will be more clear.

File: scripts/auth.ps1 (hunk 7 above)
Comment (omelianlevkovych) [reply scripts/auth.ps1]: Sure, the main suggestion behind not using double quotas everywhere stands for it possible unexpected behavior. In case you are using double quotas you are actually allowing your script to modify the literal inside the quotas with some logic. Lets make an example:
For instance, you have a urlDomain variable $urlDomain = 'https://mydomain' and in case you are using double quotas you can do the following
$urlDomain = 'https://mydomain'
"The url domain is $urlDomain" //Double quotas used. Output: The url domain is https://mydomain
However in case you do use single quota such logic would not be possible
$urlDomain = 'https://mydomain'
'The url domain is $urlDomain' //Single quotas used. Output: The url domain is $urlDomain
Also for instance you have some numeric operations inside your string
"The value of $(1+1) is 2" //Double quotas used. Output: The value of 2 is 2 
And with single quotas
'The value of $(1+1) is 2' //Single quotas used. Output: The value of $(1+1) is 2
So basically, you should have a rule - if you do not want to have any logic or variable value inside your string, do not use double quotas, and use single one instead!

File: scripts/Auth_scripts/auth.ps1 (hunk 13 above)
Comment (omelianlevkovych) [reply scripts/Auth_scripts/auth.ps1]: Actually the global variables is also not the most reliable things to have, because they provide unnecessary dependencies and one day they can break the user environment. The better approach to this is to let your function returns actually the values you are looking for (accessToken). 

someFunction {
   $x = 0
   return $x
}

$y = someFunction
In case you have to return few variables, you always can use object as return type.

File: scripts/auth.ps1 (hunk 8 above)
Comment (blevkovych) [reply scripts/auth.ps1]: I can do this but I'm still against it. In my opinion it's unnecessary. Every time user runs this script it'll ask him if he wants to update help? That's stupid innit bruv. Again in case user doesn't have administrator privileges it'll throw an error and whole script won't work because of that. After a little bit of investigation I found out there is no way to track if help files are up to date so we can't even notify user about that either.

File: scripts/Auth_scripts/main.ps1 (hunk 9 above)
Comment (blevkovych) [reply scripts/Auth_scripts/main.ps1]: In your opinion what version of mode we should use? Also starting from version 2  it "Prohibit references to non-existent properties of an object." which exist in our case if user provide bad URL endpoint and the following error message becomes misleading. Please be so kind to elaborate why do we even need to use Set-StrictMode.

File: scripts/Auth_scripts/main.ps1 (hunk 10 above)
Comment (blevkovych) [reply scripts/Auth_scripts/main.ps1]: It's hard coded because the localhost endpoint same all the time. And if user chooses local env it's always going to be this URL. But i'll rename it. And if user chooses external endpoint we are going to reassign the value.

Comment (blevkovych) [reply scripts/Auth_scripts/main.ps1]: Renamed.

File: scripts/Auth_scripts/auth.ps1 (hunk 11 above)
Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: Changed.

File: scripts/Auth_scripts/auth.ps1 (hunk 12 above)
Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: Done

File: scripts/Auth_scripts/auth.ps1 (hunk 13 above)
Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: Changed

Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: The problem with return statement is that every time you want to to anything with script you will have to authorize again and again. To avoid such behavior we actually need to store it somewhere and for that I choose environment variables, but global sounds even better.

File: scripts/auth.ps1 (hunk 8 above)
Comment (omelianlevkovych) [reply scripts/auth.ps1]: It makes sense, lets not include it into the script.

File: scripts/Auth_scripts/main.ps1 (hunk 9 above)
Comment (omelianlevkovych) [reply scripts/Auth_scripts/main.ps1]: It just does not make so much sense looking throw the script and almost in the same place there are comment and command line stating for two different things. Please, do not take comments too personal, it is just misunderstanding. Can you explain what your comment just right above states for? And what the difference between strict mode v1.0?
And now answer to your question, the cmdlet Strict-Mode command generates an error when the content of a script violates best-practices coding rules.

File: scripts/Auth_scripts/auth.ps1 [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+                continue
+            }
+            else 
+            {
+               throw
+            }
+        }
+        break
+    }
+    Set-Variable -Name "accessToken" -Scope global -Value (($response.content | ConvertFrom-Json).token).ToString()
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/auth.ps1]: please, use single quotas

File: scripts/Auth_scripts/auth.ps1 [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        $headers = @{
+            'Content-Type' = 'application/json'
+        }
+        try
+        {
+            $response = Invoke-WebRequest -Uri $urlDomain/api/account/login -Method POST -Body $body -Headers $headers
+        }
+        catch
+        {
+            if (  $_.Exception.response.StatusCode -eq "Unauthorized" )
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/auth.ps1]: please, use single quotas

File: scripts/Auth_scripts/getpet.ps1 [hunk 16]
```diff
@@ -0,0 +1,25 @@
+Function GetPet([string]$urlDomain)
+{
+    $pet_id = Read-Host 'Please enter Pet Id'
+    $headers = @{
+        Authorization = "Bearer $accessToken"
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/getpet.ps1]: please, use single quotas

File: scripts/Auth_scripts/getpet.ps1 [hunk 17]
```diff
@@ -0,0 +1,25 @@
+Function GetPet([string]$urlDomain)
+{
+    $pet_id = Read-Host 'Please enter Pet Id'
+    $headers = @{
+        Authorization = "Bearer $accessToken"
+    }
+    try
+    {
+        $response = Invoke-RestMethod -Uri $urlDomain/api/pets/$pet_id -Headers $headers -ContentType "application/json"
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/getpet.ps1]: please, use single quotas

File: scripts/Auth_scripts/getpet.ps1 [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    $headers = @{
+        Authorization = "Bearer $accessToken"
+    }
+    try
+    {
+        $response = Invoke-RestMethod -Uri $urlDomain/api/pets/$pet_id -Headers $headers -ContentType "application/json"
+    }
+    catch
+    {  
+        if (  $_.Exception.response.StatusCode -eq "Unauthorized" )
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/getpet.ps1]: please, use single quotas

File: scripts/Auth_scripts/main.ps1 [hunk 19]
```diff
@@ -0,0 +1,39 @@
+# Requires PSVersion 5.1.18362.1171
+. ".\auth.ps1"
+. ".\getpet.ps1"
+. ".\menu.ps1"
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/main.ps1]: please, use single quotas

File: scripts/Auth_scripts/main.ps1 [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+        public bool CheckValidationResult(
+            ServicePoint srvPoint, X509Certificate certificate,
+            WebRequest request, int certificateProblem) {
+            return true;
+        }
+    }
+"@
+[System.Net.ServicePointManager]::CertificatePolicy = New-Object TrustAllCertsPolicy
+
+$urlDomain = "https://localhost:5001"
```
Comment (omelianlevkovych) [inline scripts/Auth_scripts/main.ps1]: please, use single quotas

File: scripts/Auth_scripts/auth.ps1 (hunk 13 above)
Comment (omelianlevkovych) [reply scripts/Auth_scripts/auth.ps1]: Can't you store as local variable the function result in your main.ps1 and then if needed provide it to the next function as parameter?

**Comment by omelianlevkovych** [comment]
Looks great 🥇. 
Please, fix some minor things and we will hold the PR just up to the certificate changes.

File: scripts/Auth_scripts/auth.ps1 (hunk 13 above)
Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: You can, but what if you need to start the script all over again the minute later? You'll then have to authorize again. And that's every time you run main.ps1. Or do we only want save the token for current session. I mean Is it okay that user has to authorize every time he starts main.ps1?

Comment (omelianlevkovych) [reply scripts/Auth_scripts/auth.ps1]: What exactly do you mean by user session here? Each time you close PS and open it again? In this case - yes, you should input the credentials each time you close and open power shell again. Have you tried to do non single session (when you close and open PS) with global variables (not the environment variables). I presume you will not be able to do that.

File: scripts/Auth_scripts/main.ps1 (hunk 9 above)
Comment (blevkovych) [reply scripts/Auth_scripts/main.ps1]: 5.1 version stands for for PowerShell itself. As for strict mode there is 3 versions. In version 2 and 3 there is problem with  properties of an object that's non-existent. For example in our cause when we Invoke-WebRequest if client doesn't provide proper URL the error message will say that "The property 'StatusCode' cannot be found on this object. Verify that the property exists." but in version 1 It'll say "The remote name could not be resolved". I guess in the end we can use strict mode 2 and 3, but then I'll need somehow handle such errors.

File: scripts/Auth_scripts/auth.ps1 (hunk 13 above)
Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: By session I mean calling main.ps1 once, every time you call main.ps1 you'll have to input login and password to get token. We can avoid that by using environment or global variables, local variables will disappear after executing the srcipt once. And environment variables also will disappear if you close PS - FYI, actually I believe any variables.

File: scripts/Auth_scripts/getpet.ps1 (hunk 16 above)
Comment (blevkovych) [reply scripts/Auth_scripts/getpet.ps1]: Actually here I need double quotas))))

File: scripts/Auth_scripts/main.ps1 (hunk 20 above)
Comment (blevkovych) [reply scripts/Auth_scripts/main.ps1]: Check

File: scripts/Auth_scripts/main.ps1 (hunk 19 above)
Comment (blevkovych) [reply scripts/Auth_scripts/main.ps1]: done

File: scripts/Auth_scripts/getpet.ps1 (hunk 18 above)
Comment (blevkovych) [reply scripts/Auth_scripts/getpet.ps1]: boom

File: scripts/Auth_scripts/getpet.ps1 (hunk 17 above)
Comment (blevkovych) [reply scripts/Auth_scripts/getpet.ps1]: yes

File: scripts/Auth_scripts/auth.ps1 (hunk 15 above)
Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: done

File: scripts/Auth_scripts/auth.ps1 (hunk 14 above)
Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: done

File: scripts/Auth_scripts/main.ps1 (hunk 9 above)
Comment (blevkovych) [reply scripts/Auth_scripts/main.ps1]: Update: everything is good now we are using Strict-Mode v.3 and PowerShell v.5.1!

File: scripts/Auth_scripts/getpet.ps1 (hunk 16 above)
Comment (omelianlevkovych) [reply scripts/Auth_scripts/getpet.ps1]: Yep, missed it, gj

File: scripts/Auth_scripts/auth.ps1 (hunk 13 above)
Comment (omelianlevkovych) [reply scripts/Auth_scripts/auth.ps1]: There is no point to execute main script few times actually. The flow would be something like this main -> (choose what you want to do) -> fucntion -> (complete some job) - in case you want to proceed one more action it is okay to login one more time (anyway we will change the token expiration time something to 5-10 mins).

Comment (blevkovych) [reply scripts/Auth_scripts/auth.ps1]: Okay for example you ran main.ps1, got the pet data also got the token along the way. Three minutes later you run the same main.ps1 from the beginning. The Question: Do we want to have the token already (that was saved from previous interaction with main.ps1) or user should authorize again?

Comment (omelianlevkovych) [reply scripts/Auth_scripts/auth.ps1]: Yea, okay. Lets proceed with global variable here.

File: scripts/auth.ps1 (hunk 4 above)
Comment (blevkovych) [reply scripts/auth.ps1]: After small investigation, I would suggest removing this block of code because the ssl certificates will be provided during deployment to production environment and everything will work fine there. And in case you face any issue on the localhost then you should reslove this problem by adding the untrusted certificate to trusted ones.

Comment (omelianlevkovych) [reply scripts/auth.ps1]: Sounds good.

**Review by omelianlevkovych — DISMISSED** [review]
Nice job, buddy 💯

**Review by omelianlevkovych — APPROVED** [review]
